File: 2015/day01.py
import time

start = time.time()
with open("../inputs/2015/01.input", "r") as f:
    data = f.read().strip()

part1 = data.count("(") - data.count(")")

level = 0
for i, c in enumerate(data, start=1):
    level += {"(": 1, ")": -1}[c]
    if level == -1:
        part2 = i
        break


print(f"Day 01     {part1:<14} {part2:<14} {(time.time() - start)*1e3:>11.2f}")

File: rust-2020/src/day01.rs
use std::collections::HashSet;
use std::iter::FromIterator;

pub fn read(contents: &str) -> HashSet<i32> {
    HashSet::from_iter(contents.lines().map(|line| line.parse::<i32>().unwrap()))
}

pub fn part1(data: &HashSet<i32>) -> i32 {
    for &n in data {
        if data.contains(&(2020 - n)) {
            return n * (2020 - n);
        }
    }
    unreachable!();
}

pub fn part2(data: &HashSet<i32>) -> i32 {
    for &n in data {
        for &p in data {
            if data.contains(&(2020 - n - p)) {
                return n * p * (2020 - n - p);
            }
        }
    }
    unreachable!();
}

pub fn run() -> (String, String) {
    let contents = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day01.input"));
    let data = read(contents);
    let p1 = part1(&data).to_string();
    let p2 = part2(&data).to_string();
    (p1, p2)
}
